src/main/java/mate/academy/hibernate/relations/dao/impl/AbstractDao.java

```diff
@@ -1,11 +1,45 @@
 package mate.academy.hibernate.relations.dao.impl;
 
+import java.util.Optional;
+import mate.academy.hibernate.relations.exceptions.DataProcessingException;
+import org.hibernate.Session;
 import org.hibernate.SessionFactory;
+import org.hibernate.Transaction;
 
 public abstract class AbstractDao {
     protected final SessionFactory factory;
 
     protected AbstractDao(SessionFactory sessionFactory) {
         this.factory = sessionFactory;
     }
+
+    protected <T> T add(T entity) {
+        Session session = factory.openSession();
+        Transaction transaction = null;
+        try {
+            transaction = session.beginTransaction();
+            session.persist(entity);
+            transaction.commit();
+        } catch (RuntimeException e) {
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw new DataProcessingException("Failed to add "
+                    + entity.getClass().getSimpleName()
+                    + " = [" + entity + "]", e);
+        } finally {
+            session.close();
+        }
```

Comment:
Use try with resources for the `Session` object to ensure it's always closed, even when exceptions occur. You can omit the explicit `session.close()` in the `finally` block if you do this.